plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java

```diff
@@ -38,35 +44,58 @@
 public class IcebergPageSource
         implements ConnectorPageSource
 {
+    private final Type[] columnTypes;
     private final Block[] prefilledBlocks;
     private final int[] delegateIndexes;
+    private final int[] expectedColumnIndexes;
     private final ConnectorPageSource delegate;
     private final Optional<ReaderProjectionsAdapter> projectionsAdapter;
+    private final DeleteFilter<TrinoRow> deleteFilter;
 
     public IcebergPageSource(
-            List<IcebergColumnHandle> columns,
+            List<IcebergColumnHandle> expectedColumns,
+            List<IcebergColumnHandle> allColumns,
             Map<Integer, Optional<String>> partitionKeys,
             ConnectorPageSource delegate,
-            Optional<ReaderProjectionsAdapter> projectionsAdapter)
+            Optional<ReaderProjectionsAdapter> projectionsAdapter,
+            DeleteFilter<TrinoRow> deleteFilter)
     {
-        int size = requireNonNull(columns, "columns is null").size();
+        // expectedColumns should contain columns which should be in the final Page
+        // allColumns should include all expectedColumns as well as any columns needed by the DeleteFilter
+        requireNonNull(expectedColumns, "expectedColumns is null");
+        requireNonNull(allColumns, "allColumns is null");
+        this.expectedColumnIndexes = new int[expectedColumns.size()];
+        for (int i = 0; i < expectedColumns.size(); i++) {
+            expectedColumnIndexes[i] = allColumns.indexOf(expectedColumns.get(i));
```

Comment:
This is quadratic. 
Can we expect `expectedColumns` and `allColumns` be in same order?

Reply:
I need to rework this a little to handle projection pushdown, will see if I can make it linear too.